utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+		DynamicEvent(
+			self.0
+				.pointer("/event")
+				.expect("existence of /event is ensured by constructor")
+				.clone(),
+		)
+	}
+
+	/// Unwrap the inner JSON value.
+	pub fn into_inner(self) -> JsonValue {
+		self.0
+	}
+}
+
+impl DynamicEvent {
+	/// If this event represents the outcome of an extrinsic, return the result.
+	pub fn extrinsic_outcome(&self) -> Option<Result<(), DispatchError>> {
+		if self.0.pointer("/System/ExtrinsicSuccess/dispatch_info").is_some() {
+			Some(Ok(()))
+		} else {
+			self.0
+				.pointer("/System/ExtrinsicFailed/dispatch_error")
+				.map(|dispatch_error_json| {
+					Err((|| {
+						Some(DispatchError::KnownModuleError {
+							pallet: dispatch_error_json["pallet"].as_str()?.to_string(),
+							name: dispatch_error_json["name"].as_str()?.to_string(),
+							error: dispatch_error_json["error"].as_str()?.to_string(),
+						})
+					})()
+					.unwrap_or_else(|| DispatchError::DispatchError(dispatch_error_json.clone())))
+				})
+		}
+	}
+
+	pub fn to_json(self) -> JsonValue {
+		self.0.clone()
+	}
+}
+
+type NamedVariantLookup<T> = BTreeMap<u8, (String, T)>;
+type ErrorInfo = NamedVariantLookup<Vec<String>>;
+
+/// A decoder for substrate events.
+///
+/// Events are decoded dynamically based on metadata provided on construction of this type.
+///
+/// For the System.ExtrinsicFailed event, the dispatch error is resolved to a structured
+/// human-readable error message with the following shape:
+///
+/// ```json
+/// {
+///     "pallet": "PalletName",
+///     "name": "ErrorName",
+///     "error": "Human readable error message"
+/// }
+/// ```
+#[derive(Debug)]
+pub struct EventDecoder {
+	events_type_id: u32,
+	types: PortableRegistry,
+	errors: NamedVariantLookup<ErrorInfo>,
+}
+
+impl EventDecoder {
+	/// Create a new event decoder.
+	///
+	/// The arguments to this constructor can be extracted from substrate metadata.
+	pub fn new(mut types: PortableRegistry, errors_type_id: u32) -> Option<Self> {
+		let errors = match types.resolve(errors_type_id).unwrap().type_def.clone() {
+			scale_info::TypeDef::Variant(runtime_error_type) => runtime_error_type
+				.variants
+				.into_iter()
+				.map(|pallet_error_type| {
+					(
+						pallet_error_type.index,
+						(pallet_error_type.name, {
+							let type_id = pallet_error_type
+								.fields
+								.first()
+								.expect("error variant has exactly one field")
+								.ty
+								.id;
+							match types.resolve(type_id).unwrap().type_def.clone() {
+								scale_info::TypeDef::Variant(pallet_errors) => pallet_errors
+									.variants
+									.into_iter()
+									.map(|error_variant| {
+										(
+											error_variant.index,
+											(error_variant.name, error_variant.docs),
+										)
+									})
+									.collect::<ErrorInfo>(),
+								_ => panic!("Inner error type is not an Enum"),
+							}
+						}),
+					)
+				})
+				.collect::<NamedVariantLookup<_>>(),
+			_ => panic!("Outer error type is not an Enum"),
```

Comment:
nit: maybe slightly more explicit / context specific, like we expect it be an enum because we expect it to specifically be the highest level RuntimeError enum
```suggestion
			_ => panic!("Outer error type is not an Enum"),
```

Reply:
I updated the diff in your comment with an alernative error message - is that better?

Reply:
works for me